common/ledger/blkstorage/file_util.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package blkstorage
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+func createAndSyncFileAtomically(dir, tmpFile, finalFile string, content []byte) error {
+	tempFilePath := filepath.Join(dir, tmpFile)
+	finalFilePath := filepath.Join(dir, finalFile)
+	if err := createAndSyncFile(tempFilePath, content); err != nil {
+		return err
+	}
+	if err := os.Rename(tempFilePath, finalFilePath); err != nil {
+		return err
+	}
+	return syncDir(dir)
+}
+
+func createAndSyncFile(filePath string, content []byte) error {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0644)
+	if err != nil {
+		return errors.Wrapf(err, "error while creating file:%s", filePath)
+	}
+	_, err = file.Write(content)
+	if err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while writing to file:%s", filePath)
+	}
+	if err = file.Sync(); err != nil {
+		file.Close()
+		return errors.Wrapf(err, "error while synching the file:%s", filePath)
+	}
+	if err := file.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing the file:%s", filePath)
+	}
+	return nil
+}
+
+func syncDir(dirPath string) error {
+	dir, err := os.Open(dirPath)
+	if err != nil {
+		return errors.Wrapf(err, "error while opening dir:%s", dirPath)
+	}
+	if err := dir.Sync(); err != nil {
+		dir.Close()
+		return errors.Wrapf(err, "error while synching dir:%s", dirPath)
+	}
+	if err := dir.Close(); err != nil {
+		return errors.Wrapf(err, "error while closing dir:%s", dirPath)
+	}
+	return err
+}
```

Comment:
Duplicated from https://github.com/hyperledger/fabric/blob/2fdbafbed9fc1213a28e5980b8c334ce04c27dcb/core/ledger/kvledger/snapshot.go#L155

https://github.com/hyperledger/fabric/blob/2fdbafbed9fc1213a28e5980b8c334ce04c27dcb/core/ledger/kvledger/snapshot.go#L179?

If we had following the Golang project layout -- https://github.com/golang-standards/project-layout we could have kept all these utility function under the folder _pkg/file_. Anyway, this layout itself was proposed only a year ago. Just a bad luck.

Is there a way to create a pkg/file somewhere and use it across fabric?

Reply:
I was planning to move the common code related to files operation under `common/ledger/filesutil` after this PR. But we can discuss if you have a better suggestion. `common/ledger/util` also has some similar functions that can also be merged.

Reply:
Once you shared this link with me https://github.com/cockroachdb/cockroach/tree/master/pkg/util/fileutil I liked it too. So I am fine with `fileutil`. If we ever refactor the whole directory structure of Fabric, we can put everything inside pkg (https://github.com/golang-standards/project-layout/tree/master/pkg). Until then, `common/ledger/filesutil` is good.